File: src/main/java/no/nav/syfo/util/DateUtil.java
package no.nav.syfo.util;

import no.nav.syfo.domain.Periode;

import java.time.LocalDate;

public class DateUtil {

    public static boolean overlapperPerioder(Periode a, Periode b) {
        return compare(a.getFom()).isBetweenOrEqual(b)
                || compare(a.getTom()).isBetweenOrEqual(b)
                || compare(b.getFom()).isBetweenOrEqual(a)
                || compare(b.getTom()).isBetweenOrEqual(a);
    }

    public static LocalDateComparator compare(LocalDate localDate) {
        return new LocalDateComparator(localDate);
    }

    public static class LocalDateComparator {
        private LocalDate localDate;

        private LocalDateComparator(LocalDate localDate) {
            this.localDate = localDate;
        }

        public boolean isNotBetween(LocalDate fom, LocalDate tom) {
            return localDate.isBefore(fom) || localDate.isAfter(tom);
        }

        public boolean isBetweenOrEqual(Periode p) {
            return !isNotBetween(p.getFom(), p.getTom());
        }
    }
}

File: src/main/kotlin/no/nav/syfo/util/DateUtil.kt
package no.nav.syfo.util

import java.time.LocalDate
import no.nav.syfo.domain.Periode

object DateUtil {
    fun overlapperPerioder(a: Periode, b: Periode): Boolean {
        return (compare(a.fom).isBetweenOrEqual(b)
            || compare(a.tom).isBetweenOrEqual(b)
            || compare(b.fom).isBetweenOrEqual(a)
            || compare(b.tom).isBetweenOrEqual(a))
    }

    fun compare(localDate: LocalDate): LocalDateComparator {
        return LocalDateComparator(localDate)
    }

    class LocalDateComparator internal constructor(private val localDate: LocalDate) {
        private fun isNotBetween(fom: LocalDate?, tom: LocalDate?): Boolean {
            return localDate.isBefore(fom) || localDate.isAfter(tom)
        }

        fun isBetweenOrEqual(p: Periode): Boolean {
            return !isNotBetween(p.fom, p.tom)
        }
    }
}
